File: PadraoGOF/Criacionais/Prototype/Prototype.py
import copy

class Pessoa:
    def __init__(self, id, nome, idade):
        self.id = id
        self.nome = nome
        self.idade = idade

    def clone(self):
        return copy.copy(self)

    def __str__(self):
        return f'ID: {self.id} | Nome: {self.nome} | Idade: {self.idade}'

# Corrigido: A classe deve ficar fora da classe Pessoa (não aninhada)
class GerenciaPessoa:
    def __init__(self):
        self.grupopessoas = {}

    def addPessoa(self, id, nome, idade):
        pessoa = Pessoa(id, nome, idade)
        self.grupopessoas[id] = pessoa

    def getPessoa(self, id):
        return self.grupopessoas[id].clone()

gerencia = GerenciaPessoa()

# Cadastrando pessoas originais
gerencia.addPessoa(1, 'Joao da Silva', 30)
gerencia.addPessoa(2, 'Maria da Silva', 30)
gerencia.addPessoa(3, 'Jose da Silva', 30)

# Criando os clones
pessoaClone1 = gerencia.getPessoa(1)
pessoaClone2 = gerencia.getPessoa(2)
pessoaClone3 = gerencia.getPessoa(3)

# Modificando clones
pessoaClone1.nome = 'Clonaldo de Oliveira'
pessoaClone2.nome = 'Bruno Clonaldo'
pessoaClone3.idade = 99

print("====== Originais =======")
print(gerencia.getPessoa(1))
print(gerencia.getPessoa(2))
print(gerencia.getPessoa(3))

print("\n====== Clones Modificados =======")
print(pessoaClone1)
print(pessoaClone2)
print(pessoaClone3)

File: PadraoGOF/Criacionais/Prototype/Prototype.js
//classe pessoa referencia
class Pessoa{
    constructor(id, nome, idade){
        this.id = id;
        this.nome = nome;
        this.idade = idade;
    }
//metodo para clonar e realizar uma cópia
    clone(){
        return new Pessoa(this.id,this.nome, this.idade);
    }
}

//classe Gerencia Pessoa para gerenciar istanciais 

class GerenciaPessoa{
    constructor(){
        this.pessoas = {};
    }

    //add uma nova pessoa
    addPessoa(id, nome, idade){
        const pessoa = new Pessoa(id, nome, idade);
        this.pessoas[id] = pessoa;
    }
    //Solicita uma pessoa
    getPessoaByid(id){
        const pessoaOriginal = this.pessoas[id];
        if (pessoaOriginal){
            return pessoaOriginal.clone();
        }else {
            return null;
        }
    }
}

//criando uma instancia de Gerenciar pessoas 
const gerencia = new GerenciaPessoa();

//Add Peeosas
gerencia.addPessoa(1, 'Rogerin', 44)
gerencia.addPessoa(1, 'ManoMarcos', 36)
gerencia.addPessoa(3, 'Thigas', 32)

//clonando Pessoas e modificando

const pessoaClone1 = gerencia.getPessoaByid(1);
const pessoaClone3 = gerencia.getPessoaByid(3)
if (pessoaClone1){
    pessoaClone1.nome = 'Clonaldo de Oliveira';
}
if (pessoaClone3){
    pessoaClone3.nome = 'Hckeado do Santos';
    pessoaClone3.idade = "133"  

}

//imprimindo Pesssoa

console.log("====== Pessoas Originais =======")
console.log(gerencia.getPessoaByid(1));
console.log(gerencia.getPessoaByid(2));
console.log(gerencia.getPessoaByid(3));

console.log("======== Pessoas Clonadas =======")

console.log(pessoaClone1);
console.log(pessoaClone3);